### PR TITLE
Update GPU messaging benchmarks

### PR DESCRIPTION
**Pull request by minitu**
Refer to OSU latency & bandwidth benchmarks (http://mvapich.cse.ohio-state.edu/benchmarks/) and update GPU messaging benchmarks.

Also adds timing breakdowns to analyze GPU messaging performance inside Charm++.

### CODE REVIEW

File: benchmarks/charm++/cuda/gpudirect/latency/latency.C [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    main_proxy = thisProxy;
+    min_size = 1;
+    max_size = 4194304;
+    n_iters_reg = 1000;
+    n_iters_large = 100;
+    warmup_iters = 10;
+    zerocopy = false;
+
+    if (CkNumPes() != 2) {
+      CkAbort("There should be 2 PEs");
```
Comment (matthiasdiener) [inline benchmarks/charm++/cuda/gpudirect/latency/latency.C:31]:
```suggestion
      CkPrintf("Error: there should be 2 PEs");
      CkExit(1);
```

File: benchmarks/charm++/cuda/gpudirect/latency/latency.C [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+          break;
+        default:
+          CkAbort("Unknown command line argument detected");
+      }
+    }
+    delete m;
+
+    if (n_iters_reg > MAX_ITERS || n_iters_large > MAX_ITERS) {
+      CkPrintf("Number of iterations must be less than %d\n", MAX_ITERS);
+      CkExit();
```
Comment (matthiasdiener) [inline benchmarks/charm++/cuda/gpudirect/latency/latency.C:61]:
```suggestion
      CkExit(1);
```

File: benchmarks/charm++/cuda/gpudirect/latency/latency.C [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+          n_iters_reg = atoi(optarg);
+          break;
+        case 'l':
+          n_iters_large = atoi(optarg);
+          break;
+        case 'w':
+          warmup_iters = atoi(optarg);
+          break;
+        default:
+          CkAbort("Unknown command line argument detected");
```
Comment (matthiasdiener) [inline benchmarks/charm++/cuda/gpudirect/latency/latency.C:54]:
```suggestion
          CkPrintf("Unknown command line argument detected");
          CkExit(1);
```

File: benchmarks/charm++/cuda/gpudirect/bandwidth/bandwidth.C [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    min_size = 1;
+    max_size = 4194304;
+    n_iters_reg = 1000;
+    n_iters_large = 100;
+    warmup_iters = 10;
+    zerocopy = false;
+    window_size = 64;
+
+    if (CkNumPes() != 2) {
+      CkAbort("There should be 2 PEs");
```
Comment (minitu) [inline benchmarks/charm++/cuda/gpudirect/bandwidth/bandwidth.C:33]:
```suggestion
      CkPrintf("Error: there should be 2 PEs");
      CkExit(1);
```

File: benchmarks/charm++/cuda/gpudirect/bandwidth/bandwidth.C [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+          n_iters_large = atoi(optarg);
+          break;
+        case 'w':
+          warmup_iters = atoi(optarg);
+          break;
+        case 'd':
+          window_size = atoi(optarg);
+          break;
+        default:
+          CkAbort("Unknown command line argument detected");
```
Comment (minitu) [inline benchmarks/charm++/cuda/gpudirect/bandwidth/bandwidth.C:59]:
```suggestion
          CkPrintf("Unknown command line argument detected");
          CkExit(1);
```

File: benchmarks/charm++/cuda/gpudirect/bandwidth/bandwidth.C [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+      }
+    }
+    delete m;
+
+    if (n_iters_reg > MAX_ITERS || n_iters_large > MAX_ITERS) {
+      CkPrintf("Number of iterations must be less than %d\n", MAX_ITERS);
+      CkExit();
```
Comment (minitu) [inline benchmarks/charm++/cuda/gpudirect/bandwidth/bandwidth.C:66]:
```suggestion
      CkExit(1);
```

**Review by matthiasdiener — DISMISSED** [review]

**Comment by minitu** [comment]
Can I get another review for this?

File: src/ck-core/ckrdmadevice.C [hunk 7]
```diff
@@ -54,10 +54,23 @@
...
+#endif
+
 CsvExtern(GPUManager, gpu_manager);
 
 // Invoked after post entry method
 bool CkRdmaDeviceIssueRgets(envelope *env, int numops, void **arrPtrs, int *arrSizes, CkDeviceBufferPost *postStructs) {
+#if TIMING_BREAKDOWN
+  static double total_times[N_TIMER] = {0};
+  static int count = 0;
+  count++;
```
Comment (matthiasdiener) [inline src/ck-core/ckrdmadevice.C:69]: Are these safe for SMP mode?

Comment (minitu) [reply src/ck-core/ckrdmadevice.C:69]: Good point, I forgot that the timing breakdown was included in this patch. Will update it to use Cpv variables instead

**Comment by minitu** [comment]
@matthiasdiener Using thread_local to support SMP.

**Review by matthiasdiener — APPROVED** [review]

**Comment by minitu** [comment]
Can I get one more review for this?

**Pull request merged**